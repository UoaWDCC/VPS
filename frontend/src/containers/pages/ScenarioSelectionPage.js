import React, { useContext, useEffect, useState } from "react";
import { useHistory } from "react-router-dom";
import SideBar from "../../components/SideBar";
import ListContainer from "../../components/ListContainer";
import ScreenContainer from "../../components/ScreenContainer";
import ScenarioContext from "../../context/ScenarioContext";
import { usePut } from "../../hooks/crudHooks";
import AuthenticationContext from "../../context/AuthenticationContext";

/**
 * Page that shows the user's existing scenarios.
 *
 * @container
 */
export default function ScenarioSelectionPage({ data = null }) {
  const { scenarios, currentScenario, setCurrentScenario, reFetch } =
    useContext(ScenarioContext);
  const { getUserIdToken } = useContext(AuthenticationContext);
  const history = useHistory();

  // invalid name state stores the last item that had a null name, will display error message
  const [invalidNameId, setInvalidNameId] = useState("");
  /** function is called when the user unfocuses from a scenario name */
  async function changeScenarioName({ target }) {
    /**
     * if target value of name entered is empty, of of just spaces, instantly revert to the previous name
     */
    if (
      target.value === "" ||
      target.value === null ||
      target.value.trim() === ""
    ) {
      target.value = currentScenario.name;
      setInvalidNameId(currentScenario._id);
    } else {
      setInvalidNameId("");
    }

    await usePut(
      `/api/scenario/${currentScenario._id}`,
      {
        ...currentScenario,
        name: target.value,
      },
      getUserIdToken
    );
    reFetch();
  }

  /** function is called when the user double clicks a scenario */
  async function editScenario() {
    // should be set on the first click, but check to be sure anyway
    if (currentScenario != null) {
      history.push(`/scenario/${currentScenario._id}`);
    }
  }

  useEffect(() => {
    setCurrentScenario(null);
    reFetch();
  }, []);

  return (
    <ScreenContainer>
      <SideBar />
      <ListContainer
        data={data || scenarios}
        onItemSelected={setCurrentScenario}
        onItemDoubleClick={editScenario}
        onItemBlur={changeScenarioName}
        invalidNameId={invalidNameId}
      />
    </ScreenContainer>
  );
}
