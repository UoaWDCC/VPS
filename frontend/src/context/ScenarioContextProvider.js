import React, { useState } from "react";
import { useGet } from "../hooks/crudHooks";
import useLocalStorage from "../hooks/useLocalStorage";
import ScenarioContext from "./ScenarioContext";

export default function ScenarioContextProvider({ children }) {
  const [currentScenario, setCurrentScenario] = useLocalStorage(
    "currentScenario",
    null
  );
  const [scenarios, setScenarios] = useState();
  const { reFetch } = useGet(`api/scenario`, setScenarios);

  return (
    <ScenarioContext.Provider
      value={{
        scenarios,
        setScenarios,
        reFetch,
        currentScenario,
        setCurrentScenario,
      }}
    >
      {children}
    </ScenarioContext.Provider>
  );
}
