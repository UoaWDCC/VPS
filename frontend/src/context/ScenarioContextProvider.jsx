import { useState } from "react";
import { useGet } from "../hooks/crudHooks";
import useLocalStorage from "../hooks/useLocalStorage";
import ScenarioContext from "./ScenarioContext";

/**
 * This is a Context Provider made with the React Context API
 * ScenarioContextProvider allows access to scenario info and the refetch function
 */
export default function ScenarioContextProvider({ children }) {
  const [currentScenario, setCurrentScenario] = useLocalStorage(
    "currentScenario",
    null
  );
  const [scenarios, setScenarios] = useState();
  const [assignedScenarios, setAssignedScenarios] = useState();
  const [roleList, setRoleList] = useState();

  const { reFetch } = useGet(`api/scenario`, setScenarios);
  const { reFetch: reFetch2 } = useGet(
    `api/scenario/assigned`,
    setAssignedScenarios
  );

  const { reFetch: reFetch3 } = useGet(
    currentScenario ? `api/group/${currentScenario._id}/roleList` : null,
    currentScenario ? setRoleList : () => {}
  );

  return (
    <ScenarioContext.Provider
      value={{
        scenarios,
        setScenarios,
        reFetch,
        assignedScenarios,
        setAssignedScenarios,
        reFetch2,
        reFetch3,
        currentScenario,
        setCurrentScenario,
        roleList,
      }}
    >
      {children}
    </ScenarioContext.Provider>
  );
}
