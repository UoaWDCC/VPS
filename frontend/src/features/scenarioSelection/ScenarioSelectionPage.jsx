import MenuItem from "@material-ui/core/MenuItem";
import { useContext, useEffect, useState } from "react";
import { useHistory } from "react-router-dom";
import ContextMenu from "../../components/ContextMenu";
import ThumbnailList from "../../components/ListContainer/ThumbnailList";
import ScreenContainer from "../../components/ScreenContainer/ScreenContainer";
import SideBar from "../../components/SideBar/SideBar";
import AuthenticationContext from "../../context/AuthenticationContext";
import ScenarioContext from "../../context/ScenarioContext";
import AccessLevel from "../../enums/route.access.level";
import { useDelete, usePut } from "../../hooks/crudHooks";

import MovieFilterRoundedIcon from "@mui/icons-material/MovieFilterRounded";
import TheatersRoundedIcon from "@mui/icons-material/TheatersRounded";

/**
 * Page that shows the user's existing scenarios.
 *
 * @container
 */
export default function ScenarioSelectionPage() {
  const {
    scenarios: userScenarios,
    reFetch,
    assignedScenarios,
    reFetch2,
    currentScenario,
    setCurrentScenario,
  } = useContext(ScenarioContext);
  const { getUserIdToken, VpsUser } = useContext(AuthenticationContext);
  const history = useHistory();

  /** Handle right-click to open up ContextMenu */
  const [contextMenuPosition, setContextMenuPosition] = useState(null);
  function handleContextMenu(event) {
    event.preventDefault();
    if (contextMenuPosition == null) {
      setContextMenuPosition({ x: event.clientX, y: event.clientY });
    } else {
      setContextMenuPosition(null);
    }
  }

  /** Calls backend end point to switch to the lecturer's dashboard */
  function openDashboard() {
    history.push(`/dashboard/${currentScenario._id}`);
  }

  /** Calls backend end point to delete a scenario. */
  async function deleteScenario() {
    await useDelete(`/api/scenario/${currentScenario._id}`, getUserIdToken);
    setCurrentScenario(null);
    reFetch();
  }

  /** Opens new window to play selected scenario. */
  function playScenario() {
    window.open(`/play/${currentScenario._id}`, "_blank");
  }

  // invalid name state stores the last item that had a null name, will display error message
  const [invalidNameId, setInvalidNameId] = useState("");

  /** function is called when the user unfocuses from a scenario name */
  async function changeScenarioName({ target }) {
    /**
     * if target value of name entered is empty, of of just spaces, instantly revert to the previous name
     */
    if (
      target.value === "" ||
      target.value === null ||
      target.value.trim() === ""
    ) {
      target.value = currentScenario.name;
      setInvalidNameId(currentScenario._id);
    } else {
      setInvalidNameId("");
    }

    await usePut(
      `/api/scenario/${currentScenario._id}`,
      {
        ...currentScenario,
        name: target.value,
      },
      getUserIdToken
    );
    reFetch();
  }

  /** function is called when the user double clicks a scenario */
  async function editScenario() {
    // should be set on the first click, but check to be sure anyway
    if (currentScenario != null) {
      history.push(`/scenario/${currentScenario._id}`);
    }
  }

  useEffect(() => {
    setCurrentScenario(null);
    reFetch();

    reFetch2();
  }, []);

  return (
    <ScreenContainer>
      <SideBar />
      <div onContextMenu={handleContextMenu} className="w-full h-full">
        <ContextMenu
          position={contextMenuPosition}
          setPosition={setContextMenuPosition}
        >
          <MenuItem onClick={playScenario} disabled={!currentScenario}>
            Play
          </MenuItem>
          <MenuItem onClick={editScenario} disabled={!currentScenario}>
            Edit
          </MenuItem>
          <MenuItem onClick={deleteScenario} disabled={!currentScenario}>
            Delete
          </MenuItem>
          {VpsUser.role === AccessLevel.STAFF ? (
            <MenuItem onClick={openDashboard} disabled={!currentScenario}>
              Dashboard
            </MenuItem>
          ) : (
            <MenuItem onClick={openDashboard} disabled={!currentScenario}>
              Dashboard
            </MenuItem>
          )}
        </ContextMenu>

        {/* Scenario List */}
        <div className="w-full h-full px-10 py-10 overflow-x-hidden overflow-y-scroll flex flex-col gap-10">
          {/* List of scenarios created by the logged-in user */}
          {userScenarios && (
            <div>
              <h1 className="text-3xl font-mona font-bold my-3 flex items-center gap-3">
                <MovieFilterRoundedIcon fontSize="large" /> Your Scenarios
              </h1>

              <div>
                <ThumbnailList
                  // data={userScenarios}
                  data={userScenarios.map((scenario) => {
                    scenario.components = scenario.thumbnail?.components || [];
                    return scenario;
                  })}
                  onItemSelected={setCurrentScenario}
                  onItemDoubleClick={editScenario}
                  onItemBlur={changeScenarioName}
                  invalidNameId={invalidNameId}
                />
              </div>
            </div>
          )}

          {/* List of scenarios assigned to the logged-in user */}
          {assignedScenarios && (
            <div>
              <h1 className="text-3xl font-mona font-bold my-3 flex items-center gap-3">
                <TheatersRoundedIcon fontSize="large" /> Assigned Scenarios
              </h1>
              <ThumbnailList
                data={assignedScenarios.map((scenario) => {
                  scenario.components = scenario.thumbnail?.components || [];
                  return scenario;
                })}
                onItemSelected={(scenario) => {
                  // For assigned scenarios, play the scenario on click.
                  window.open(`/play/${scenario._id}`, "_blank");
                }}
                invalidNameId={invalidNameId}
                highlightOnSelect={false}
              />
            </div>
          )}
        </div>
      </div>
    </ScreenContainer>
  );
}
