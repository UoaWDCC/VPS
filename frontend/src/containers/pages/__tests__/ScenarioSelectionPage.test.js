import React from "react";
import renderer from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";
import ScenarioSelectionPage from "../ScenarioSelectionPage";
import ScenarioContextProvider from "../../../context/ScenarioContextProvider";

const dummyScenarios = [
  { id: 1, name: "test1" },
  { id: 2, name: "test2" },
];

test("Scenario Selection page snapshot test", () => {
  const component = renderer.create(
    <BrowserRouter>
      <ScenarioContextProvider>
        <ScenarioSelectionPage data={dummyScenarios} />
      </ScenarioContextProvider>
    </BrowserRouter>
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
