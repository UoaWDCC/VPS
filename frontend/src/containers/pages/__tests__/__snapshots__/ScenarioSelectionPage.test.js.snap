// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scenario Selection page snapshot test 1`] = `
<div
  className="rowContainer"
>
  <div
    className="sideBar"
  >
    <img
      alt="UoA Medical & Health Science Logo"
      className="logo"
      src="uoa-med-and-health-sci-logo.png"
    />
    <ul
      className="sideBarList"
    >
      <li>
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained btn side contained white"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            Create
          </span>
        </button>
      </li>
      <li>
        <a
          aria-disabled={true}
          className="MuiButtonBase-root MuiButton-root MuiButton-contained btn side contained white buttonDisabled   Mui-disabled Mui-disabled"
          href="/scenario/null"
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex={-1}
        >
          <span
            className="MuiButton-label"
          >
            Edit
          </span>
        </a>
      </li>
    </ul>
  </div>
  <div
    className="scenarioListContainer"
  >
    <ul
      className="MuiImageList-root"
      style={
        Object {
          "margin": -15,
        }
      }
    >
      <li
        className="MuiImageListItem-root"
        height={200}
        onClick={[Function]}
        style={
          Object {
            "height": 240,
            "padding": 15,
            "width": "25%",
          }
        }
      >
        <div
          className="MuiImageListItem-item"
        >
          <div
            className="imageListItem"
          >
            <div
              className="MuiBox-root MuiBox-root-1"
            />
            <p
              className="text"
            >
              test1
            </p>
          </div>
        </div>
      </li>
      <li
        className="MuiImageListItem-root"
        height={200}
        onClick={[Function]}
        style={
          Object {
            "height": 240,
            "padding": 15,
            "width": "25%",
          }
        }
      >
        <div
          className="MuiImageListItem-item"
        >
          <div
            className="imageListItem"
          >
            <div
              className="MuiBox-root MuiBox-root-2"
            />
            <p
              className="text"
            >
              test2
            </p>
          </div>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
